Added ability Parasitic Waste

miriamlefae opened the pull request:
## Summary
Ported the radical red ability Parasitic Waste and confirmed it works as it should. Added the ability for Garbodor-Mega.

## Description

### Ability Implementation
- Defined a new ability constant `ABILITY_PARASITIC_WASTE` and added the ability.

- Added a new function `CanTriggerParasiticWaste`
    - Returns true if move has a secondary poison chance
    - Returns true for max moves that can poison

- Updated case `MOVEEND_ABSORB` in Cmd_moveend to check for `CanTriggerParasiticWaste`
    - Added a conditional to explicitly set damage healed to half if `CanTriggerParasiticWaste` is true

- Added checks to prevent poison if Parasitic Waste is triggered
  - Updated case `STATUS1_POISON` in `SetMoveEffect` to prevent regular poison.
  - Updated case `STATUS1_TOXIC_POISON` to prevent toxic poison as well.

- Reworded the RR description to be more clear about what the the ability does.

## Other tasks
- Added Parasitic Waste to ability randomiser whitelist
- Changed Mega Garbodor's ability to Parasitic Waste


Note: Dire Claw is intentionally left out to maintain parity with radical red

## Images
![parasiticwaste_1](https://github.com/user-attachments/assets/78fbfb2b-f395-418f-8e85-7c354b92d830) ![parasiticwaste_2](https://github.com/user-attachments/assets/9c87c7c8-691c-429d-83a3-b671ea35179e) ![parasiticwaste_3](https://github.com/user-attachments/assets/1026958f-f379-4021-a3f2-d07132facfa4)

![parsiticwaste_bulletproof](https://github.com/user-attachments/assets/3e7e0fa8-fcbb-4891-b03d-fab42cea4f44) ![parsiticwaste_pb_charge](https://github.com/user-attachments/assets/80dedd60-1978-4d2c-9d27-7bd02d946277) ![parsiticwaste_steel](https://github.com/user-attachments/assets/943ec207-daee-438b-9645-c2f95d36ca0d) ![pw_badly_poisons_without](https://github.com/user-attachments/assets/bcf263d4-51cb-4ce1-a1e0-4fce492b248b) ![pw_no_badly_poison](https://github.com/user-attachments/assets/2b9aad7b-d732-4fc4-bf7b-b75d672ca120) ![pw_poisons_ifnot](https://github.com/user-attachments/assets/ff6d4be6-7063-443d-a500-e96c5a3793a8)

#### Summary screen
<img width="240" height="160" alt="pokeemerald-0" src="https://github.com/user-attachments/assets/d9c7f1f1-baca-4c36-977c-49a416dbb2b8" />



## **People who collaborated with me in this PR**
@RagnarDev27 (aka ScarredLion)

## **Discord contact info**
kildemal

miriamlefae committed:
feat: add ability Parasitic waste

- Defined a new ability constant and added the ability.

- Added a new function CanTriggerParasiticWaste
    - Returns true if move has a secondary poison chance
    - Returns true for max moves that can poison

- Updated case MOVEEND_ABSORB in Cmd_moveend to check for CanTriggerParasiticWaste
    - Added a conditional to explicitly set damage healed to half if CanTriggerParasiticWaste is true

- Updated case STATUS1_POISON in SetMoveEffect to prevent poison if Parasitic Waste is triggered.

- Reworded the RR description to be more clear about what the the ability does.

Note: Dire Claw is intentionally left out to maintain parity with radical red

iriv24 commented on src/battle_util.c:
```diff
@@ -11639,6 +11639,34 @@ u32 GetBattlerMoveTargetType(u32 battler, u32 move)
+        MoveHasAdditionalEffect(move, MOVE_EFFECT_TOXIC);
+
+    bool32 IsParasiticWasteTriggered =
+        IsParasiticWasteEffect && gMovesInfo[move].effect == EFFECT_HIT;
```
i dont believe you need this line. not sure if this was intentional, but the only thing this line does is explicitly exclude shell side arm and barb barrage. if those moves and those moves only are supposed to be excluded, then this is fine for now.

miriamlefae replied:
You're right. That _is_ redundant and it breaks barb and shell side arm lol. Changing

miriamlefae replied:
done

iriv24 commented on src/battle_util.c:
```diff
@@ -11639,6 +11639,34 @@ u32 GetBattlerMoveTargetType(u32 battler, u32 move)
+    bool32 IsParasiticWasteTriggered =
+        IsParasiticWasteEffect && gMovesInfo[move].effect == EFFECT_HIT;
+
+    bool32 IsParasticWasteTriggeredMax =
```
i don think you need this either, as all the moves that meet the criteria for `IsParasiticWasteMaxEffect` will also be guaranteed to meet this condition so it is redundant

miriamlefae replied:
done

iriv24 commented:
looks good so far, just a few comments. other things to address:
- can you add parasitic waste to the end of `src/data/randomizer/ability_whitelist.h` so it is included in the randomizer
- can you add it to mega garbo in its species info and just replace Stench in every slot for it
- can you test this with Sludge Wave in a double battle? i _think_ it would work like Parabolic Charge but im curious (would like to see a video of this one)

the following cases dont need a video, just things i would check if i were implementing this and would like to get at least verbal confirmation that they work:
- regression test using a parasitic waste-affected move as mega garbo into a steel-type and make sure nothing weird happens
- regression test using sludge bomb into a bulletproof mon as mega garbo and make sure nothing weird happens
- regression test that moves that can poison can still poison and behave normally when the user does _not_ have parasitic waste

iriv24 commented on src/battle_script_commands.c:
```diff
@@ -3139,6 +3139,9 @@ void SetMoveEffect(bool32 primary, bool32 certain)
 
+            if (CanTriggerParasiticWaste(gBattleScripting.battler, gCurrentMove))
+                break;
+
```
i think youll also need this block under `STATUS1_TOXIC_POISON`, you can test it with malignant chain or poison fang

miriamlefae replied:
Oh yeah, missed that

miriamlefae replied:
I did check that moves poison when the user does not have parasitic waste. Will check the rest.

miriamlefae replied:
all done. gifs attached

miriamlefae committed:
chore: add Parasitic Waste to randomiser whitelist

miriamlefae committed:
chore: add parasitic waste to mega garbodor

miriamlefae committed:
fix: remove redundant code that broke moves

    - Shell Side Arm and Barb Barrage wouldn't work due to the check for EFFECT_HIT
    - Removed redundant check for EFFECT_MAX_MOVE

miriamlefae committed:
fix: fix parasitic waste allowing toxic poison

    - Parasitic Waste now prevents toxic poisoning if triggered

iriv24 commented on src/battle_util.c:
```diff
@@ -11639,6 +11639,29 @@ u32 GetBattlerMoveTargetType(u32 battler, u32 move)
+        gMovesInfo[move].argument == MAX_EFFECT_POISON_FOES;
+
+    bool32 IsParasiticWasteEffect =
+        IsParasiticWasteMaxEffect ||
```
at this point it probably makes the most sense to just remove `IsParasiticWasteMaxEffect` as a variable and put its checks under `IsParasiticWasteEffect` directly

miriamlefae replied:
done  7eea73f1fe7bc444a1c90cd8201f143589b02b4f

iriv24 commented on src/data/abilities.h:
```diff
@@ -2524,6 +2524,13 @@ const struct Ability gAbilitiesInfo[ABILITIES_COUNT] =
+    [ABILITY_PARASITIC_WASTE] =
+    {
+        .name = _("Parasitic Waste"),
+        .description = COMPOUND_STRING("Attacks that poison drain instead."),
```
last ask is to make sure this fits on the summary screen and doesn't need any new lines or anything!

miriamlefae replied:
done. image attached

miriamlefae committed:
refactor: combine vars in parasitic waste func

- Remove IsParasiticWasteEffectMax and incorporate it into IsParasiticWasteEffect
